### PR TITLE
Cylc.rose pre release refactor

### PR DESCRIPTION
**Pull request by wxtim**
A generic tidy up of Cylc Rose, attempting to comment the code better and remove the odd unnecessary line.
Should not contain any functional changes.

Has [accompanying cylc-flow PR](https://github.com/cylc/cylc-flow/pull/4097)

- variable name change: dir_ => srcdir
- variable name change: dest_root => destdir
- Entry points to have just one top level function each.
- Top level functions ordered by order in which they are called.
- Move checks for input srcdir and destdir being filepaths or None
  to top level entry point functions to reduce duplication.
- Remove support for inputs to top level functions being str
- Tested path checking
- Re-read edited docstrings.
- Tidied up utils.py
- Removed rose-dump-log function from fileinstall to top level of the
  entry point becasue it's unrelated to fileinstall.

### CODE REVIEW

File: tests/test_generic_utils.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        ([Path('/cycle/path/'), None], [Path('/cycle/path'), None]),
+        # Different starting types re-typed independently:
+        (
+            [Path('/cycle/path/'), '/bridle/path'],
+            [Path('/cycle/path'), Path('/bridle/path')]),
+    ]
+)
+def test_paths_to_pathlib(paths, expect):
+    assert paths_to_pathlib(paths) == expect
+
```
Comment (datamel) [inline tests/test_generic_utils.py:43]:
```suggestion
```
Extra line is causing pycodestyle to fail for me.

**Review by datamel — APPROVED** [review]
Looks good to me, thanks Tim. Just need to resolve the conflicts and fix the pycodestyle.

**Review by oliver-sanders — APPROVED** [review]

**Comment by datamel** [comment]
I've now got pytest failures on this branch when checked out with: https://github.com/cylc/cylc-flow/pull/4097

**Comment by datamel** [comment]
> I've now got pytest failures on this branch when checked out with: [cylc/cylc-flow#4097](https://github.com/cylc/cylc-flow/pull/4097)

This was me being daft and had items in my global config from another branch.

**Pull request merged**